pulsar-client-kafka-compat/src/main/java/org/apache/kafka/clients/consumer/PulsarKafkaConsumer.java

```diff
@@ -0,0 +1,482 @@
+        throw new UnsupportedOperationException("ConsumerRebalanceListener is not supported");
+    }
+
+    @Override
+    public void assign(Collection<TopicPartition> partitions) {
+        throw new UnsupportedOperationException("Cannot manually assign partitions");
+    }
+
+    @Override
+    public void subscribe(Pattern pattern, ConsumerRebalanceListener callback) {
+        throw new UnsupportedOperationException("Cannot subscribe with topic name pattern");
+    }
+
+    @Override
+    public void unsubscribe() {
+        consumers.values().forEach(c -> {
+            try {
+                c.unsubscribe();
+            } catch (PulsarClientException e) {
+                throw new RuntimeException(e);
+            }
+        });
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public ConsumerRecords<K, V> poll(long timeoutMillis) {
+        try {
+            QueueItem item = receivedMessages.poll(timeoutMillis, TimeUnit.MILLISECONDS);
+            if (item == null) {
+                return (ConsumerRecords<K, V>) ConsumerRecords.EMPTY;
+            }
+
+            if (isAutoCommit) {
+                // Commit the offset of previously dequeued messages
+                commitAsync();
+            }
+
+            String topic = item.consumer.getTopic();
+            Message msg = item.message;
+            MessageIdImpl msgId = (MessageIdImpl) msg.getMessageId();
+            long offset = getOffset(msgId);
+
+            int partition = msgId.getPartitionIndex();
+
+            TopicPartition tp = new TopicPartition(topic, partition);
+
+            K key = getKey(topic, msg);
+            V value = valueDeserializer.deserialize(topic, msg.getData());
+
+            ConsumerRecord<K, V> consumerRecord = new ConsumerRecord<>(topic, partition, offset, msg.getPublishTime(),
+                    TimestampType.CREATE_TIME, -1, msg.hasKey() ? msg.getKey().length() : 0, msg.getData().length, key,
+                    value);
+
+            Map<TopicPartition, List<ConsumerRecord<K, V>>> records = new HashMap<>();
+            records.put(tp, Lists.newArrayList(consumerRecord));
+
+            // Update last message id seen by application
+            lastMessageId.put(item.consumer, msgId);
+            return new ConsumerRecords<>(records);
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private K getKey(String topic, Message msg) {
+        if (!msg.hasKey()) {
+            return null;
+        }
+
+        if (keyDeserializer instanceof StringDeserializer) {
+            return (K) msg.getKey();
+        } else {
+            // Assume base64 encoding
+            byte[] data = Base64.getDecoder().decode(msg.getKey());
+            return keyDeserializer.deserialize(topic, data);
+        }
+    }
+
+    @Override
+    public void commitSync() {
```

Comment:
can commitSync be a shortcut of commitAsync and wait for callback?

Reply:
I'll try to refactor it. as it is `commitAsync()` doesn't return or take a callback, I'll add a private method and use it from the API methods.